File: sip-phone-lib/src/main/java/com/telcobright/sipphone/verto/VertoClient.java
package com.telcobright.sipphone.verto;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.UUID;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Verto WebSocket client for FreeSWITCH signaling.
 * Pure Java — no Android dependencies.
 */
public class VertoClient {

    private static final Logger log = LoggerFactory.getLogger(VertoClient.class);

    public interface VertoEventListener {
        void onConnected();
        void onDisconnected(String reason);
        void onLoginSuccess(String sessionId);
        void onLoginFailed(String error);
        void onIncomingCall(String callId, String callerNumber, String sdp);
        void onCallAnswered(String callId, String sdp);
        void onCallEnded(String callId, String reason);
        void onMediaUpdate(String callId, String sdp);
        void onError(String error);
    }

    private final String serverUrl;
    private final String login;
    private final String password;
    private final VertoEventListener listener;

    private final Gson gson = new Gson();
    private final OkHttpClient client;
    private WebSocket webSocket;
    private final AtomicInteger requestId = new AtomicInteger(1);
    private final ConcurrentHashMap<Integer, CompletableFuture<JsonObject>> pendingRequests = new ConcurrentHashMap<>();
    private String sessionId;
    private final ExecutorService executor = Executors.newSingleThreadExecutor(r -> {
        Thread t = new Thread(r, "verto-login");
        t.setDaemon(true);
        return t;
    });

    public VertoClient(String serverUrl, String login, String password, VertoEventListener listener) {
        this.serverUrl = serverUrl;
        this.login = login;
        this.password = password;
        this.listener = listener;
        this.client = new OkHttpClient.Builder()
                .pingInterval(30, TimeUnit.SECONDS)
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .build();
    }

    public void connect() {
        Request request = new Request.Builder().url(serverUrl).build();
        webSocket = client.newWebSocket(request, new WebSocketListener() {
            @Override public void onOpen(WebSocket ws, Response response) {
                log.info("WebSocket connected to {}", serverUrl);
                listener.onConnected();
                performLogin();
            }

            @Override public void onMessage(WebSocket ws, String text) {
                handleMessage(text);
            }

            @Override public void onClosing(WebSocket ws, int code, String reason) {
                ws.close(1000, null);
            }

            @Override public void onClosed(WebSocket ws, int code, String reason) {
                log.warn("WebSocket CLOSED — server={}, code={}, reason='{}'", serverUrl, code, reason);
                listener.onDisconnected(reason);
            }

            @Override public void onFailure(WebSocket ws, Throwable t, Response response) {
                String detail = t.getMessage() != null ? t.getMessage() : "Unknown error";
                int httpCode = response != null ? response.code() : -1;
                log.error("WebSocket FAILED — server={}, error='{}', httpCode={}", serverUrl, detail, httpCode);
                listener.onDisconnected(detail);
            }
        });
    }

    public void disconnect() {
        if (webSocket != null) {
            webSocket.close(1000, "User disconnect");
            webSocket = null;
        }
        executor.shutdownNow();
    }

    public String invite(String destination, String localSdp) {
        String callId = UUID.randomUUID().toString();
        JsonObject params = new JsonObject();
        params.addProperty("callID", callId);
        params.addProperty("sdp", localSdp);
        JsonObject dialogParams = new JsonObject();
        dialogParams.addProperty("callID", callId);
        dialogParams.addProperty("destination_number", destination);
        dialogParams.addProperty("caller_id_name", login);
        dialogParams.addProperty("caller_id_number", login);
        dialogParams.addProperty("remote_caller_id_name", destination);
        dialogParams.addProperty("remote_caller_id_number", destination);
        params.add("dialogParams", dialogParams);
        sendRequest("verto.invite", params);
        return callId;
    }

    public void answer(String callId, String localSdp) {
        JsonObject params = new JsonObject();
        params.addProperty("callID", callId);
        params.addProperty("sdp", localSdp);
        JsonObject dp = new JsonObject();
        dp.addProperty("callID", callId);
        params.add("dialogParams", dp);
        sendRequest("verto.answer", params);
    }

    public void bye(String callId) {
        JsonObject params = new JsonObject();
        params.addProperty("callID", callId);
        JsonObject dp = new JsonObject();
        dp.addProperty("callID", callId);
        params.add("dialogParams", dp);
        sendRequest("verto.bye", params);
    }

    public void sendDtmf(String callId, String digits) {
        JsonObject params = new JsonObject();
        params.addProperty("callID", callId);
        params.addProperty("dtmf", digits);
        JsonObject dp = new JsonObject();
        dp.addProperty("callID", callId);
        params.add("dialogParams", dp);
        sendRequest("verto.info", params);
    }

    private void performLogin() {
        sessionId = UUID.randomUUID().toString();
        JsonObject params = new JsonObject();
        params.addProperty("login", login);
        params.addProperty("passwd", password);
        params.addProperty("sessid", sessionId);

        executor.submit(() -> {
            try {
                log.info("Sending login request for user '{}' to {}", login, serverUrl);
                JsonObject response = sendRequestAsync("login", params).get(10, TimeUnit.SECONDS);
                if (response.has("sessid")) {
                    sessionId = response.get("sessid").getAsString();
                    log.info("Registration SUCCESS — user='{}', sessionId={}", login, sessionId);
                    listener.onLoginSuccess(sessionId);
                } else {
                    log.warn("Registration FAILED — user='{}', response: {}", login, response);
                    listener.onLoginFailed("Login failed — unexpected response");
                }
            } catch (java.util.concurrent.TimeoutException e) {
                log.error("Registration TIMEOUT — user='{}', server={} (no response in 10s)", login, serverUrl);
                listener.onLoginFailed("Login timeout — no response from server");
            } catch (java.util.concurrent.ExecutionException e) {
                String cause = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
                log.error("Registration REJECTED — user='{}', server={}, error: {}", login, serverUrl, cause);
                listener.onLoginFailed("Login rejected: " + cause);
            } catch (Exception e) {
                log.error("Registration ERROR — user='{}', server={}: {}", login, serverUrl, e.getMessage(), e);
                listener.onLoginFailed(e.getMessage() != null ? e.getMessage() : "Login error");
            }
        });
    }

    private void handleMessage(String text) {
        try {
            JsonObject json = JsonParser.parseString(text).getAsJsonObject();

            if (json.has("id")) {
                int id = json.get("id").getAsInt();
                CompletableFuture<JsonObject> future = pendingRequests.remove(id);
                if (future != null) {
                    if (json.has("result")) {
                        future.complete(json.getAsJsonObject("result"));
                    } else if (json.has("error")) {
                        future.completeExceptionally(new Exception(json.getAsJsonObject("error").toString()));
                    }
                    return;
                }
            }

            if (json.has("method")) {
                handleEvent(json);
            }
        } catch (Exception e) {
            log.error("Failed to parse message: {}", e.getMessage());
        }
    }

    private void handleEvent(JsonObject json) {
        String method = json.get("method").getAsString();
        JsonObject params = json.has("params") ? json.getAsJsonObject("params") : null;
        if (params == null) return;
        int eventId = json.has("id") ? json.get("id").getAsInt() : 0;

        switch (method) {
            case "verto.invite" -> {
                String callId = params.has("callID") ? params.get("callID").getAsString() : null;
                if (callId == null) return;
                String sdp = params.has("sdp") ? params.get("sdp").getAsString() : "";
                String caller = "Unknown";
                if (params.has("dialogParams")) {
                    var dp = params.getAsJsonObject("dialogParams");
                    if (dp.has("caller_id_number")) caller = dp.get("caller_id_number").getAsString();
                }
                listener.onIncomingCall(callId, caller, sdp);
            }
            case "verto.answer" -> {
                String callId = params.has("callID") ? params.get("callID").getAsString() : null;
                if (callId == null) return;
                String sdp = params.has("sdp") ? params.get("sdp").getAsString() : "";
                listener.onCallAnswered(callId, sdp);
                sendResult(eventId, new JsonObject());
            }
            case "verto.bye" -> {
                String callId = params.has("callID") ? params.get("callID").getAsString() : null;
                if (callId == null) return;
                String reason = params.has("cause") ? params.get("cause").getAsString() : "Normal";
                listener.onCallEnded(callId, reason);
                sendResult(eventId, new JsonObject());
            }
            case "verto.media" -> {
                String callId = params.has("callID") ? params.get("callID").getAsString() : null;
                if (callId == null) return;
                String sdp = params.has("sdp") ? params.get("sdp").getAsString() : "";
                listener.onMediaUpdate(callId, sdp);
                sendResult(eventId, new JsonObject());
            }
            case "verto.display" -> sendResult(eventId, new JsonObject());
            default -> log.debug("Unhandled event: {}", method);
        }
    }

    private void sendRequest(String method, JsonObject params) {
        int id = requestId.getAndIncrement();
        JsonObject msg = new JsonObject();
        msg.addProperty("jsonrpc", "2.0");
        msg.addProperty("id", id);
        msg.addProperty("method", method);
        msg.add("params", params);
        String text = gson.toJson(msg);
        log.debug("Sending: {} (id={})", method, id);
        if (webSocket != null) webSocket.send(text);
    }

    private CompletableFuture<JsonObject> sendRequestAsync(String method, JsonObject params) {
        int id = requestId.getAndIncrement();
        CompletableFuture<JsonObject> future = new CompletableFuture<>();
        pendingRequests.put(id, future);

        JsonObject msg = new JsonObject();
        msg.addProperty("jsonrpc", "2.0");
        msg.addProperty("id", id);
        msg.addProperty("method", method);
        msg.add("params", params);
        String text = gson.toJson(msg);
        log.debug("Sending async: {} (id={})", method, id);
        if (webSocket != null) webSocket.send(text);
        return future;
    }

    private void sendResult(int id, JsonObject result) {
        if (id == 0) return;
        JsonObject msg = new JsonObject();
        msg.addProperty("jsonrpc", "2.0");
        msg.addProperty("id", id);
        msg.add("result", result);
        if (webSocket != null) webSocket.send(gson.toJson(msg));
    }
}

File: app/src/main/java/com/telcobright/sipphone/verto/VertoClient.kt
package com.telcobright.sipphone.verto

import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.*
import okhttp3.*
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * Verto WebSocket client for FreeSWITCH signaling.
 *
 * Implements the Verto JSON-RPC protocol over WebSocket:
 * - login (authentication)
 * - verto.invite (outbound call)
 * - verto.answer (accept inbound call)
 * - verto.bye (hang up)
 * - verto.media (SDP exchange)
 */
class VertoClient(
    private val serverUrl: String,
    private val login: String,
    private val password: String,
    private val listener: VertoEventListener
) {

    companion object {
        private const val TAG = "VertoClient"
        private const val PING_INTERVAL_SEC = 30L
    }

    interface VertoEventListener {
        fun onConnected()
        fun onDisconnected(reason: String)
        fun onLoginSuccess(sessionId: String)
        fun onLoginFailed(error: String)
        fun onIncomingCall(callId: String, callerNumber: String, sdp: String)
        fun onCallAnswered(callId: String, sdp: String)
        fun onCallEnded(callId: String, reason: String)
        fun onMediaUpdate(callId: String, sdp: String)
        fun onError(error: String)
    }

    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .pingInterval(PING_INTERVAL_SEC, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS) // No timeout for WebSocket
        .build()

    private var webSocket: WebSocket? = null
    private val requestId = AtomicInteger(1)
    private val pendingRequests = ConcurrentHashMap<Int, CompletableDeferred<JsonObject>>()
    private var sessionId: String? = null
    private var scope: CoroutineScope? = null

    fun connect() {
        scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

        val request = Request.Builder()
            .url(serverUrl)
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connected")
                listener.onConnected()
                performLogin()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closing: $code $reason")
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed: $code $reason")
                listener.onDisconnected(reason)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket failure: ${t.message}")
                listener.onDisconnected(t.message ?: "Connection failed")
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "User disconnect")
        webSocket = null
        scope?.cancel()
        scope = null
    }

    /**
     * Make an outbound call.
     *
     * @param destination  Phone number or extension
     * @param localSdp     Our SDP offer (with AMR codec)
     * @return Call ID
     */
    fun invite(destination: String, localSdp: String): String {
        val callId = UUID.randomUUID().toString()

        val params = JsonObject().apply {
            addProperty("callID", callId)
            addProperty("sdp", localSdp)
            add("dialogParams", JsonObject().apply {
                addProperty("callID", callId)
                addProperty("destination_number", destination)
                addProperty("caller_id_name", login)
                addProperty("caller_id_number", login)
                addProperty("remote_caller_id_name", destination)
                addProperty("remote_caller_id_number", destination)
            })
        }

        sendRequest("verto.invite", params)
        return callId
    }

    /**
     * Answer an incoming call.
     *
     * @param callId    Call ID from onIncomingCall
     * @param localSdp  Our SDP answer (with AMR codec)
     */
    fun answer(callId: String, localSdp: String) {
        val params = JsonObject().apply {
            addProperty("callID", callId)
            addProperty("sdp", localSdp)
            add("dialogParams", JsonObject().apply {
                addProperty("callID", callId)
            })
        }

        sendRequest("verto.answer", params)
    }

    /**
     * Hang up a call.
     */
    fun bye(callId: String) {
        val params = JsonObject().apply {
            addProperty("callID", callId)
            add("dialogParams", JsonObject().apply {
                addProperty("callID", callId)
            })
        }

        sendRequest("verto.bye", params)
    }

    /**
     * Send DTMF digits.
     */
    fun sendDtmf(callId: String, digits: String) {
        val params = JsonObject().apply {
            addProperty("callID", callId)
            addProperty("dtmf", digits)
            add("dialogParams", JsonObject().apply {
                addProperty("callID", callId)
            })
        }

        sendRequest("verto.info", params)
    }

    private fun performLogin() {
        sessionId = UUID.randomUUID().toString()

        val params = JsonObject().apply {
            addProperty("login", login)
            addProperty("passwd", password)
            addProperty("sessid", sessionId)
        }

        scope?.launch {
            try {
                val response = sendRequestAsync("login", params)
                if (response.has("sessid")) {
                    sessionId = response.get("sessid").asString
                    listener.onLoginSuccess(sessionId!!)
                } else {
                    listener.onLoginFailed("Login failed")
                }
            } catch (e: Exception) {
                listener.onLoginFailed(e.message ?: "Login error")
            }
        }
    }

    private fun handleMessage(text: String) {
        try {
            val json = JsonParser.parseString(text).asJsonObject

            /* Check if this is a response to a pending request */
            if (json.has("id")) {
                val id = json.get("id").asInt
                val deferred = pendingRequests.remove(id)
                if (deferred != null) {
                    if (json.has("result")) {
                        deferred.complete(json.getAsJsonObject("result"))
                    } else if (json.has("error")) {
                        deferred.completeExceptionally(
                            Exception(json.getAsJsonObject("error").toString())
                        )
                    }
                    return
                }
            }

            /* This is an inbound event from FreeSWITCH */
            if (json.has("method")) {
                handleEvent(json)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse message: ${e.message}")
        }
    }

    private fun handleEvent(json: JsonObject) {
        val method = json.get("method").asString
        val params = json.getAsJsonObject("params") ?: return
        val eventId = if (json.has("id")) json.get("id").asInt else 0

        when (method) {
            "verto.invite" -> {
                val callId = params.get("callID")?.asString ?: return
                val sdp = params.get("sdp")?.asString ?: ""
                val callerNumber = params.getAsJsonObject("dialogParams")
                    ?.get("caller_id_number")?.asString ?: "Unknown"
                listener.onIncomingCall(callId, callerNumber, sdp)
            }

            "verto.answer" -> {
                val callId = params.get("callID")?.asString ?: return
                val sdp = params.get("sdp")?.asString ?: ""
                listener.onCallAnswered(callId, sdp)
                sendResult(eventId, JsonObject())
            }

            "verto.bye" -> {
                val callId = params.get("callID")?.asString ?: return
                val reason = params.get("cause")?.asString ?: "Normal"
                listener.onCallEnded(callId, reason)
                sendResult(eventId, JsonObject())
            }

            "verto.media" -> {
                val callId = params.get("callID")?.asString ?: return
                val sdp = params.get("sdp")?.asString ?: ""
                listener.onMediaUpdate(callId, sdp)
                sendResult(eventId, JsonObject())
            }

            "verto.display" -> {
                /* Display update — just acknowledge */
                sendResult(eventId, JsonObject())
            }

            else -> {
                Log.d(TAG, "Unhandled event: $method")
            }
        }
    }

    private fun sendRequest(method: String, params: JsonObject) {
        val id = requestId.getAndIncrement()
        val msg = JsonObject().apply {
            addProperty("jsonrpc", "2.0")
            addProperty("id", id)
            addProperty("method", method)
            add("params", params)
        }

        val text = gson.toJson(msg)
        Log.d(TAG, "Sending: $method (id=$id)")
        webSocket?.send(text)
    }

    private suspend fun sendRequestAsync(method: String, params: JsonObject): JsonObject {
        val id = requestId.getAndIncrement()
        val deferred = CompletableDeferred<JsonObject>()
        pendingRequests[id] = deferred

        val msg = JsonObject().apply {
            addProperty("jsonrpc", "2.0")
            addProperty("id", id)
            addProperty("method", method)
            add("params", params)
        }

        val text = gson.toJson(msg)
        Log.d(TAG, "Sending async: $method (id=$id)")
        webSocket?.send(text)

        return withTimeout(10_000) { deferred.await() }
    }

    private fun sendResult(id: Int, result: JsonObject) {
        if (id == 0) return
        val msg = JsonObject().apply {
            addProperty("jsonrpc", "2.0")
            addProperty("id", id)
            add("result", result)
        }
        webSocket?.send(gson.toJson(msg))
    }
}
